doc/tn5250.1
'\" t
.ig
Man page for tn5250.

Copyright (C) 1999 - 2008 Carey Evans.

You can redistribute and/or modify this document under the terms of 
the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option)
any later version.

This document is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
..
.TH TN5250 1 "12 November 2001"
.SH NAME
tn5250 \- an implementation of the 5250 telnet protocol
.SH SYNOPSIS
.B tn5250
.RI [\| OPTIONS \|]
.I HOSTNAME
.br
.B xt5250
.RI [\| OPTIONS \|]
.I HOSTNAME
.br
.B tn5250 \-version
.SH "DESCRIPTION"
.B tn5250
emulates IBM's 5250 compatible terminals to connect over TCP/IP to an
IBM AS/400.  The terminal emulation works on any local terminal
supported by curses, and provides 132x27 and full colour support where
possible.
.PP
An IBM 5250 terminal supports a number of special keys, and does a lot 
of processing itself before sending data to the host system.  Because
a standard ASCII terminal does not support all these keys,
.B tn5250
uses control sequences to perform the operations.
.PP
Named session support (with the
.B env.DEVNAME
option) is available for most versions of OS/400, but is only supplied
with V4R3 and later versions.  For V3R2, V3R7, V4R1 and V4R2, please
see informational APAR II10918.
.SH OPTIONS
.B tn5250
options are described in the
.BR tn5250rc (5)
man page.
.SH USAGE
This manual assumes that the user is familiar with a real 5250
terminal, or another 5250 emulator, and only describes features
relevant to
.B tn5250
itself.
.SS "Keyboard Mapping"
The following table lists the 5250 functions implemented by
.BR tn5250 ,
and the corresponding keypresses.  Keys are represented as Emacs does:
.B C-a
means hold
.B Ctrl
and press
.BR A ,
.B M-a
means press
.B Esc
or
.B C-g
followed by
.BR A ,
and
.B C-M-a
means press
.B Esc
or
.B C-g
followed by
.BR C-a .
Most setups also let you use the
.B Alt
or
.B Meta
key for
.B M-
keypresses.
.PP
.TS
tab (@);
l l
__
lB l.
Function@Keypress
F1 - F10@f1 to f10, M-1 to M-
F11@f11 [1], M--
F12@f12 [1], M-=
F13 - F24@f13 to f24 [1], M-! to M-+
Enter@return, enter, C-j, C-m
Left@left
Right@right
Up@up
Down@down
Roll Up@next, pagedown, C-d, C-f
Roll Down@prev, pageup, C-b, C-u
Backspace@backspace [1]
Home@home, C-o
End@end
Insert@insert, M-i, M-delete
Delete@delete [1]
Reset@C-r, M-r
Print@C-p, M-p
Help@M-h
SysReq@C-c, M-s
Clear@M-c
FieldExit@C-k, M-x
TestReq@C-t
Toggle@M-t
Erase@C-e
Attn@C-a, M-a
Dup@M-d
Field+@C-x, + [2]
Field-@M-m, - [2]
NewLine@C-M-j
Next Field@tab, C-i
Prev Field@backtab [1]
_
Toggle mouse@C-v
Refresh@C-l, M-l
Quit@C-q
.TE
.IP 1. 3
Which keys generate f11-f24, backtab, backspace and delete is very
dependent on the configuration of the terminal.
.IP 2.
.B +
and
.B -
only work as Field+ and Field- in signed numeric fields.
.LP
.SS Display
The screen display tries to be the same as a colour 5250 terminal.
There are some unavoidable differences though:
.IP \(bu 3
ASCII terminals (and curses) do not support the 5250 column
separators.
.IP \(bu
Some terminals (mostly ones based on the VGA text mode) do not support
underlining.  If this is the case and
.B tn5250
does not detect it automatically, you should pass the
.B \-u
option.
.IP \(bu
Some terminals (such as
.BR xterm )
do not support blinking text.
.SH EXAMPLES
These are some examples of the use of
.BR tn5250 :
.TP
.I "tn5250 as400sys"
Connect to the system
.IR as400sys ,
using the default translation map (37)
and a generated session name
.RI (QPADEV #### ).
.TP
.I "tn5250 map=870 env.DEVNAME=session1 env.TERM=IBM-3477-FC as400sys"
Connect to
.I as400sys
using the translation map for CCSID 870 (Eastern Europe), the session
name SESSION1 and with a terminal supporting colour and a 132x27
display.
.SH BUGS
Please report any bugs you find to https://github.com/tn5250/tn5250/issues
.SH "SEE ALSO"
.BR tn5250rc (5),
.BR lp5250d (1),
.BR https://tn5250.github.io/ ,
.BR RFC1205 ,
.BR RFC2877 ,
.BR II10918 .
.SH COPYRIGHT
.B tn5250
is copyright
.if t \(co
1997 \- 2008 Michael Madore.  This manpage is copyright
.if t \(co
1999 \- 2008 Carey Evans.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.
.PP
You should have received a copy of the GNU Lesser General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.SH AUTHORS
.B tn5250
was written by Michael Madore, Jay Felice, Scott Klement
and others; see the AUTHORS file for details.
This manual page was written by Carey Evans.


NEWS
tn5250 NEWS
------------------------------------------------------------------------------

Wed May 04, 2005:
	* The tn5250 project is hosted on sourceforge at:
		http://tn5250.sourceforge.net
	* There is a mailing list and archives at David Gibbs midrange site:
		http://www.midrange.com
	  The mailing list is at:
		http://lists.midrange.com/mailman/listinfo/linux5250
	* There is an X11 front end written by James Rich.  The site is:
		http://www.chowhouse.com/~james/x5250
	* For all other news please visit the project page on sourceforge.

Mon Dec 06, 1999:
	* A FAQ is now in existence, thanks to Philippe Petrinko.  You can
	  browse the FAQ at http://perso.libertysurf.fr/plinux/tn5250-faq.html
	* The CVS server is now serving up the mailing list archives, thanks
	  to David Gibbs.   You can browse the archives at
	  http://cvs.cleveland.lug.net/mail-archives/tn5250



curses/tn5250.c
/*
 * Copyright (C) 1997-2008 Michael Madore
 *
 * This file is part of TN5250.
 *
 * TN5250 is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * TN5250 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "tn5250-private.h"
#include "cursesterm.h"

Tn5250Session* sess = NULL;
Tn5250Stream* stream = NULL;
Tn5250Terminal* term = NULL;
Tn5250Display* display = NULL;
Tn5250Config* config = NULL;
Tn5250Macro* macro = NULL;

/* FIXME: This should be moved into session.[ch] or something. */
static struct valid_term {
    char* name;
    char* descr;
} valid_terms[] = {
    // clang-format off
    // DBCS Terminals not yet supported.
    // { "IBM-5555-C01", "DBCS color" },
    // { "IBM-5555-B01", "DBCS monocrome" },
    { "IBM-3477-FC", "27x132 color"      },
    { "IBM-3477-FG", "27x132 monochrome" },
    { "IBM-3180-2",  "27x132 monochrome" },
    { "IBM-3179-2",  "24x80 color"       },
    { "IBM-3196-A1", "24x80 monochrome"  },
    { "IBM-5292-2",  "24x80 color"       },
    { "IBM-5291-1",  "24x80 monochrome"  },
    { "IBM-5251-11", "24x80 monochrome"  },
    { NULL,          NULL                }
    // clang-format on
};

static void syntax(void);

int main(int argc, char* argv[]) {
#ifndef NDEBUG
    const char* tracefile;
#endif

#ifdef HAVE_SETLOCALE
    setlocale(LC_ALL, "");
#endif

    config = tn5250_config_new();
    if (tn5250_config_load_default(config) == -1) {
        tn5250_config_unref(config);
        exit(1);
    }

    if (tn5250_config_parse_argv(config, argc, argv) == -1) {
        tn5250_config_unref(config);
        syntax();
    }

    if (tn5250_config_get(config, "help")) {
        syntax();
    }
    else if (tn5250_config_get(config, "version")) {
        printf("tn5250 %s\n", VERSION);
        exit(0);
    }
    else if (!tn5250_config_get(config, "host")) {
        syntax();
    }

#ifndef NDEBUG
    tracefile = tn5250_config_get(config, "trace");
    if (tracefile != NULL) {
        tn5250_log_open(tn5250_config_get(config, "trace"));
    }
#endif

    stream = tn5250_stream_open(tn5250_config_get(config, "host"), config);
    if (stream == NULL) {
        goto bomb_out;
    }

    display = tn5250_display_new();
    if (tn5250_display_config(display, config) == -1) {
        goto bomb_out;
    }

    term = tn5250_curses_terminal_new();
    if (tn5250_config_get(config, "underscores")) {
        tn5250_curses_terminal_use_underscores(
            term, tn5250_config_get_bool(config, "underscores"));
    }
    if (tn5250_config_get(config, "ruler")) {
        tn5250_curses_terminal_display_ruler(
            term, tn5250_config_get_bool(config, "ruler"));
    }
    if ((tn5250_config_get(config, "font_80")) &&
        (tn5250_config_get(config, "font_132"))) {
        tn5250_curses_terminal_set_xterm_font(
            term, tn5250_config_get(config, "font_80"),
            tn5250_config_get(config, "font_132"));
    }
    tn5250_curses_terminal_load_colorlist(config);

    if (term == NULL) {
        goto bomb_out;
    }
    if (tn5250_terminal_config(term, config) == -1) {
        goto bomb_out;
    }
#ifndef NDEBUG
    /* Shrink-wrap the terminal with the debug terminal, if appropriate. */
    {
        const char* remotehost = tn5250_config_get(config, "host");
        if (strlen(remotehost) >= 6 && !memcmp(remotehost, "debug:", 6)) {
            Tn5250Terminal* dbgterm = tn5250_debug_terminal_new(term, stream);
            if (dbgterm == NULL) {
                tn5250_terminal_destroy(term);
                goto bomb_out;
            }
            term = dbgterm;
            if (tn5250_terminal_config(term, config) == -1) {
                goto bomb_out;
            }
        }
    }
#endif
    tn5250_terminal_init(term);
    tn5250_display_set_terminal(display, term);

    sess = tn5250_session_new();
    tn5250_display_set_session(display, sess);

    term->conn_fd = tn5250_stream_socket_handle(stream);
    tn5250_session_set_stream(sess, stream);
    if (tn5250_session_config(sess, config) == -1) {
        goto bomb_out;
    }

    macro = tn5250_macro_init();
    tn5250_macro_attach(display, macro);

    tn5250_session_main_loop(sess);
    errno = 0;

bomb_out:
    if (errno != 0) {
        printf("Could not start session: %s\n", strerror(errno));
    }

    if (macro != NULL) {
        tn5250_macro_exit(macro);
    }
    if (term != NULL) {
        tn5250_terminal_term(term);
    }
    if (sess != NULL) {
        tn5250_session_destroy(sess);
    }
    else if (stream != NULL) {
        tn5250_stream_destroy(stream);
    }
    if (config != NULL) {
        tn5250_config_unref(config);
    }
#ifndef NDEBUG
    tn5250_log_close();
#endif
    exit(0);
}

static void syntax() {
    struct valid_term* p;
    Tn5250CharMap* m;
    int i = 0;

    printf("tn5250 - TCP/IP 5250 emulator\n\
Syntax:\n\
  tn5250 [options] HOST[:PORT]\n");
#ifdef HAVE_LIBSSL
    printf("\
   To connect using ssl prefix HOST with 'ssl:'.  Example:\
      tn5250 +ssl_verify_server ssl:as400.example.com\n");
#endif
    printf("\n\
Options:\n\
   map=NAME                Character map (default is '37'):");
    m = tn5250_transmaps;
    while (m->name != NULL) {
        if (i % 5 == 0) {
            printf("\n                             ");
        }
        printf("%s, ", m->name);
        m++;
        i++;
    }
    printf("\n\
   env.DEVNAME=NAME         Use NAME as session name (default: none).\n");
#ifndef NDEBUG
    printf("\
   trace=FILE              Log session to FILE.\n");
#endif
#ifdef HAVE_LIBSSL
    printf("\
   +/-ssl_verify_server    Verify/don't verify the server's SSL certificate\n\
   ssl_ca_file=FILE        Use certificate authority (CA) certs from FILE\n\
   ssl_cert_file=FILE      File containing SSL certificate in PEM format to\n\
                           use if the AS/400 requires client authentication.\n\
   ssl_pem_pass=PHRASE     Passphrase to use when decrypting a PEM private\n\
                           key.  Used in conjunction with ssl_cert_file\n\
   ssl_check_exp[=SECS]    Check if SSL certificate is expired, or if it\n\
                           will be expired in SECS seconds.\n");
#endif
    printf("\
   +/-underscores          Use/don't use underscores instead of underline\n\
                           attribute.\n\
   +/-ruler		   Draw a ruler pointing to the cursor position\n\
   +/-version              Show emulator version and exit.\n\
   env.NAME=VALUE          Set telnet environment string NAME to VALUE.\n\
   env.TERM=TYPE           Emulate IBM terminal type (default: depends)");
    p = valid_terms;
    while (p->name) {
        printf("\n                             %s (%s)", p->name, p->descr);
        p++;
    }
    printf("\n\n");
    exit(255);
}
